packages/@react-aria/overlays/src/calculatePosition.ts

```diff
@@ -149,17 +152,24 @@ function getDelta(
   // is portaled somewhere other than the body and has an ancestor with
   // position: relative/absolute, it will be different.
   containerDimensions: Dimensions,
-  padding: number
+  padding: number,
+  containerOffsetWithBoundary: Offset
 ) {
   let containerScroll = containerDimensions.scroll[axis];
-  let boundaryHeight = boundaryDimensions[AXIS_SIZE[axis]];
-  let startEdgeOffset = offset - padding - containerScroll;
-  let endEdgeOffset = offset + padding - containerScroll + size;
-
-  if (startEdgeOffset < 0) {
-    return -startEdgeOffset;
-  } else if (endEdgeOffset > boundaryHeight) {
-    return Math.max(boundaryHeight - endEdgeOffset, -startEdgeOffset);
+  // The height/width of the boundary. Matches the axis along which we are adjusting the overlay position
+  let boundarySize = boundaryDimensions[AXIS_SIZE[axis]];
+  // Calculate the edges of the boundary (accomodating for the boundary padding) and the edges of the overlay.
+  let boundaryStartEdge = boundaryDimensions.scroll[AXIS[axis]] + padding;
+  let boundaryEndEdge = boundarySize + boundaryDimensions.scroll[AXIS[axis]] - padding;
+  let startEdgeOffset = offset - containerScroll + containerOffsetWithBoundary[axis];
+  let endEdgeOffset = offset - containerScroll + size + containerOffsetWithBoundary[axis];
+
+  // If any of the overlay edges falls outside of the boundary, shift the overlay the required amount required to align one of the overlay's
+  // edges with the closest boundary edge.
+  if (startEdgeOffset < boundaryStartEdge) {
+    return boundaryStartEdge - startEdgeOffset;
+  } else if (endEdgeOffset > boundaryEndEdge) {
+    return Math.max(boundaryEndEdge - endEdgeOffset, boundaryStartEdge - startEdgeOffset);
```

Comment:
Note that the second calculation here will prefer the minimum amount of movement of the overlay since the ifs will ensure the compared movement values are both negative (aka would shift the overlay "upwards"). I didn't add the same consideration to the first calculation because we don't have the same guarantee that the movement values are both in the same direction. Could do some calculations to ensure that but at the end of the day `getMaxHeight` will ensure that the overlay fits in the boundary anyways so it isn't really necessary anyways